src/sedpack/io/dataset_iteration.py

```diff
@@ -683,34 +685,131 @@ def as_numpy_iterator_rust(  # pylint: disable=too-many-arguments
                 f"The compression {self.dataset_structure.compression} is not "
                 "among the supported compressions: {supported_compressions}")
 
-        shard_paths: list[str] = list(
-            self._as_numpy_common(
+        with RustGenerator(
+                dataset=self,
                 split=split,
+                process_record=process_record,
                 shards=shards,
                 shard_filter=shard_filter,
-                repeat=False,
+                repeat=repeat,
+                file_parallelism=file_parallelism,
                 shuffle=shuffle,
-            ))
+        ) as rust_generator:
+            yield from rust_generator()
+
+
+class RustGenerator:
+    """A generator for tf.data.Dataset.from_generator which is reentrant (even
+    when the iteration did not finished, which can happen when using
+    tf.data.Dataset.from_generator).
+    """
+
+    def __init__(self,
+                 *,
+                 dataset: DatasetIteration,
+                 split: SplitT,
+                 process_record: None | Callable[[ExampleT], T] = None,
+                 shards: None | int = None,
+                 shard_filter: None | Callable[[ShardInfo], bool] = None,
+                 repeat: bool = True,
+                 file_parallelism: int = os.cpu_count() or 1,
+                 shuffle: int = 1_000) -> None:
+        """A reentrant generator.
+
+        Args:
+
+          dataset (DatasetIteration): The dataset being iterated.
+
+          split (SplitT): The split to be iterated.
+
+          process_record (None | Callable[[ExampleT], T]): Optional
+          transformation of each example.
+
+          shards (None | int): Optional limit on the number of used shards.
+
+          shard_filter (None | Callable[[ShardInfo], bool]): Optional predicate
+          returning True for each shard which should be iterated.
+
+          repeat (bool): Cycle infinitely.
+
+          file_parallelism (int): How many files to read in parallel.
+
+          shuffle (int): Size of the shuffle buffer.
+        """
+        self._rust_iter: None | _sedpack_rs.RustIter = None
```

Comment:
No strong preference. I tend to think that I want to know first the type the variable should have, so `T | None`.
I don't mind having the 2nd option provided we're consistent throughout the code base.
Unless we need support for older python versions, I would stay away from the 3rd option because at some point, `Optional[]` will start being deprecated.